File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/utils/DateUtils.kt
package io.mosip.vercred.vcverifier.utils

import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.DATE_REGEX
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_CURRENT_DATE_BEFORE_ISSUANCE_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_CURRENT_DATE_BEFORE_VALID_FROM
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CURRENT_DATE_BEFORE_ISSUANCE_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CURRENT_DATE_BEFORE_VALID_FROM
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_EXPIRATION_DATE_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_ISSUANCE_DATE_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_VALID_FROM_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_VALID_UNTIL_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.EXPIRATION_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ISSUANCE_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.VALID_FROM
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.VALID_UNTIL
import io.mosip.vercred.vcverifier.exception.ValidationException
import org.json.JSONObject
import org.threeten.bp.Instant
import org.threeten.bp.LocalDateTime
import org.threeten.bp.OffsetDateTime
import org.threeten.bp.ZoneOffset
import org.threeten.bp.format.DateTimeFormatter
import java.util.Date
import java.util.logging.Logger

object DateUtils {

    private val logger = Logger.getLogger(DateUtils::class.java.name)

    private val formatterWithOffset: DateTimeFormatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME
    private val formatterLocal: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME

    fun isValidDate(dateString: String): Boolean {
        return parseDate(dateString) != null
    }

    fun parseDate(dateString: String): Date? {
        return try {
            val offsetDateTime = OffsetDateTime.parse(dateString, formatterWithOffset)
            Date(offsetDateTime.toInstant().toEpochMilli())
        } catch (e: Exception) {
            try {
                val localDateTime = LocalDateTime.parse(dateString, formatterLocal)
                Date(localDateTime.toInstant(ZoneOffset.UTC).toEpochMilli())
            } catch (_: Exception) {
                null
            }
        }
    }

    fun validateV1DateFields(vcJsonObject: JSONObject) {
        listOf(
            ISSUANCE_DATE to ERROR_ISSUANCE_DATE_INVALID,
            EXPIRATION_DATE to ERROR_EXPIRATION_DATE_INVALID
        ).map { (dateKey, errorMessage) ->
            if (vcJsonObject.has(dateKey) && !isValidDate(vcJsonObject.get(dateKey).toString())) {
                throw ValidationException(errorMessage, "${ERROR_CODE_INVALID}${dateKey.uppercase()}")
            }
        }


        val issuanceDate = vcJsonObject.optString(ISSUANCE_DATE) ?: ""

        if (isFutureDateWithTolerance(issuanceDate)) {
            throw ValidationException(ERROR_CURRENT_DATE_BEFORE_ISSUANCE_DATE,
                ERROR_CODE_CURRENT_DATE_BEFORE_ISSUANCE_DATE
            )
        }

    }

    fun validateV2DateFields(vcJsonObject: JSONObject) {

        listOf(
            VALID_FROM to ERROR_VALID_FROM_INVALID,
            VALID_UNTIL to ERROR_VALID_UNTIL_INVALID
        ).map { (dateKey, errorMessage) ->
            if (vcJsonObject.has(dateKey) && !isValidDate(vcJsonObject.get(dateKey).toString())) {
                throw ValidationException(errorMessage,"${ERROR_CODE_INVALID}${dateKey.uppercase()}")
            }
        }

        if (vcJsonObject.has(VALID_FROM) && isFutureDateWithTolerance(
                vcJsonObject.optString(
                    VALID_FROM
                )
            )
        ) {
            throw ValidationException(ERROR_CURRENT_DATE_BEFORE_VALID_FROM, ERROR_CODE_CURRENT_DATE_BEFORE_VALID_FROM)
        }
    }

    fun isVCExpired(inputDate: String): Boolean {
        return inputDate.isNotEmpty() && !isFutureDateWithTolerance(inputDate)
    }

    fun isFutureDateWithTolerance(inputDateString: String, toleranceInMilliSeconds: Long = 3000): Boolean {
        val inputDate: Date? = try {
            parseDate(inputDateString)
        } catch (e: Exception) {
            logger.severe("Given date is not available in supported date formats")
            return false
        }
        if (inputDate == null) {
            logger.severe("Failed to parse the input date")
            return false
        }
        val currentTime = System.currentTimeMillis()
        val inputDateTime = inputDate.time

        val upperBound = currentTime + toleranceInMilliSeconds
        return inputDateTime > upperBound
    }

    fun formatEpochSecondsToIsoUtc(epochSeconds: Long): String {
        val odt = OffsetDateTime.ofInstant(Instant.ofEpochSecond(epochSeconds), ZoneOffset.UTC)
        return odt.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
    }

}

File: vc-verifier/js/src/validator/DateUtils.js
import {Errors, Fields} from "../constant/ValidatorConstants.js";
import {ValidationError} from "./ValidationError.js";


export const validateV1SpecificDateFields = (credential) => {
    const dateFields = [Fields.ISSUANCE_DATE, Fields.EXPIRATION_DATE]
    const errorMessage = {
        [Fields.ISSUANCE_DATE] : Errors.ERROR_ISSUANCE_DATE_INVALID,
        [Fields.EXPIRATION_DATE]: Errors.ERROR_EXPIRATION_DATE_INVALID
    }

    for (let i = 0; i < dateFields.length; i++) {
        const field = dateFields[i];
        if (credential.hasOwnProperty(field) && !isValidDate(credential[field])) {
            throw new ValidationError(errorMessage[field])
        }
    }

    const issuanceDate = new Date(credential[Fields.ISSUANCE_DATE])
    if(!isDatePassedCurrentDate(issuanceDate)){
        throw new ValidationError(`${Errors.ERROR_CURRENT_DATE_BEFORE_ISSUANCE_DATE}`)
    }

}


export const validateV2SpecificDateFields = (credential) => {
    const dateFields = [Fields.VALID_FROM, Fields.VALID_UNTIL]
    const errorMessage = {
        [Fields.VALID_FROM] : Errors.ERROR_VALID_FROM_INVALID,
        [Fields.VALID_UNTIL]: Errors.ERROR_VALID_UNTIL_INVALID
    }

    for (let i = 0; i < dateFields.length; i++) {
        const field = dateFields[i];
        if (credential.hasOwnProperty(field) && !isValidDate(credential[field])) {
            throw new ValidationError(errorMessage[field])
        }
    }

    const validFrom = new Date(credential[Fields.VALID_FROM])
    if(!isDatePassedCurrentDate(validFrom)){
        throw new ValidationError(`${Errors.ERROR_CURRENT_DATE_BEFORE_VALID_FROM}`)
    }

}

export const isVCExpired = (inputDate) => {
    return inputDate.trim() !== '' && isDatePassedCurrentDate(inputDate);
}


export const isValidDate = (dateValue) => {
    const dateFormatRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
    return dateFormatRegex.test(dateValue)
}

export const  isDatePassedCurrentDate = (inputDateString) => {
    try {
        const inputDate = new Date(inputDateString);

        if (isNaN(inputDate)) {
            return false;
        }

        const currentDate = new Date();
        return inputDate < currentDate;
    } catch (e) {
        return false;
    }
}